### PR TITLE
ARROW-5218: [C++] Improve build when third-party library locations are specified 

### PR DESCRIPTION
**Pull request by majetideepak**
Changes:
1) When searching for third-party libraries:
Look at only `#_ROOT` location if specified else check 1) `pkg_check_modules` and 2) check all other CMake paths

2) Build Uriparser only when Flight is built.
3) Find Backtrace only when specified.

### CODE REVIEW

**Comment by wesm** [comment]
@xhochy are you available to look at this in the next few days?

**Comment by majetideepak** [comment]
Failures seem unrelated to this change.

**Comment by wesm** [comment]
rebased

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -132,10 +132,13 @@ set(ARROW_SRCS
     util/task-group.cc
     util/thread-pool.cc
     util/trie.cc
-    util/uri.cc
     util/utf8.cc
     vendored/datetime/tz.cpp)
 
+if(ARROW_WITH_URIPARSER)
+  set(ARROW_SRCS ${ARROW_SRCS} util/uri.cc)
+endif()
```
Comment (wesm) [inline cpp/src/arrow/CMakeLists.txt]: uriparser is not a large dependency -- making it optional could grow onerous as time goes on, but we can see

**Review by wesm — APPROVED** [review]
+1, will merge when build runs

File: cpp/src/arrow/CMakeLists.txt (hunk 1 above)
Comment (majetideepak) [reply cpp/src/arrow/CMakeLists.txt]: Where is the Uri Parser being used currently? I couldn't find any usage.

**Comment by majetideepak** [comment]
@wesm Can we merge this?

**Comment by wesm** [comment]
@majetideepak yes, sorry about that, merging it now

**Comment by wesm** [comment]
(since I had +1'd feel free to merge in such cases when I forget)

**Comment by majetideepak** [comment]
Will do next time. Thanks!

**Pull request closed without merging**